database/src/ops/block.rs

```diff
@@ -1,89 +1,321 @@
+            add_key_image(tables.key_images_mut(), output.key.as_bytes())?;
 
-/// TODO
-pub fn get_block_long_term_weight() {
-    todo!()
-}
+            // Pre-RingCT outputs.
+            if let Some(amount) = output.amount {
+                add_output(
+                    amount,
+                    &Output {
+                        key: *output.key.as_bytes(),
+                        height: todo!(),
+                        output_flags: todo!(),
+                        tx_idx: todo!(),
+                    },
+                    tables.outputs_mut(),
+                    tables.num_outputs_mut(),
+                )?;
+            // RingCT outputs.
+            } else {
+                add_rct_output(
+                    &RctOutput {
+                        key: todo!(),
+                        height: todo!(),
+                        output_flags: todo!(),
+                        tx_idx: todo!(),
+                        commitment: todo!(),
+                    },
+                    tables.rct_outputs_mut(),
+                )?;
+            }
+        }
+    }
 
-/// TODO
-pub fn get_block_timestamp() {
-    todo!()
+    Ok(())
 }
 
-/// TODO
-pub fn get_block_cumulative_rct_outputs() {
-    todo!()
-}
+//---------------------------------------------------------------------------------------------------- `pop_block_*`
+/// Remove the top/latest block from the database.
+///
+/// # Example
+/// ```rust
+/// # use cuprate_database::{*, tables::*, ops::block::*};
+/// // TODO
+/// ```
+#[doc = doc_error!()]
+#[inline]
+pub fn pop_block(tables: &mut impl TablesMut) -> Result<BlockHeight, RuntimeError> {
+    // Remove block data from tables.
+    //
+    // TODO: What table to pop from here?
+    // Start with v3, if thats empty, try v2, etc?
+    //
+    // Branch depending on `height` and known hard fork points?
+    let (block_height, block_hash) = if todo!() {
+        let (block_height, block_info) = tables.block_info_v1s_mut().pop_last()?;
+        (block_height, block_info.block_hash)
+    } else if todo!() {
+        let (block_height, block_info) = tables.block_info_v2s_mut().pop_last()?;
+        (block_height, block_info.block_hash)
+    } else {
+        let (block_height, block_info) = tables.block_info_v3s_mut().pop_last()?;
+        (block_height, block_info.block_hash)
+    };
 
-/// TODO
-pub fn get_block() {
-    todo!()
+    // Block blobs.
+    tables.block_blobs_mut().delete(&block_height)?;
+
+    // Block heights.
+    tables.block_heights_mut().delete(&block_hash)?;
+
+    // Transaction & Outputs.
+    for () in /* block.txs */ std::iter::empty::<()>() {
+        let tx: &Transaction = todo!();
+        let tx_hash: &TxHash = todo!();
+
+        remove_tx(
+            tx_hash,
+            tables.tx_ids_mut(),
+            tables.tx_heights_mut(),
+            tables.tx_unlock_time_mut(),
+            tables.prunable_hashes_mut(),
+            tables.prunable_tx_blobs_mut(),
+        )?;
```

Comment:
When popping blocks, our `BlockInfoV*` types don't have lists of containing tx's, how do we remove them?

Loop over our `TxId => BlockHeight` table and remove until the block height is lower than the current?